pkg/virt-controller/watch/vmi.go

```diff
@@ -1866,26 +1935,25 @@ func (c *VMIController) handleHotplugVolumes(hotplugVolumes []*virtv1.Volume, ho
 		}
 		readyHotplugVolumes = append(readyHotplugVolumes, volume)
 	}
-	// Determine if the ready volumes have changed compared to the current pod
-	currentPod, oldPods := c.getActiveAndOldAttachmentPods(readyHotplugVolumes, hotplugAttachmentPods)
 
-	if currentPod == nil && len(readyHotplugVolumes) > 0 {
-		// ready volumes have changed
-		// Create new attachment pod that holds all the ready volumes
-		if err := c.createAttachmentPod(vmi, virtLauncherPod, readyHotplugVolumes); err != nil {
-			return err
-		}
-	}
+	currentPod, oldPods := c.getCurrentAndOldAttachmentPods(readyHotplugVolumes, hotplugAttachmentPods)
 
-	if len(readyHotplugVolumes) == 0 || (currentPod != nil && currentPod.Status.Phase == k8sv1.PodRunning) {
-		// Delete old attachment pod
-		for _, attachmentPod := range oldPods {
-			if err := c.deleteAttachmentPodForVolume(vmi, attachmentPod); err != nil {
-				return &syncErrorImpl{fmt.Errorf("Error deleting attachment pod %v", err), FailedDeletePodReason}
+	if len(readyHotplugVolumes) > 0 {
+		key, err := controller.KeyFunc(vmi)
+		if err != nil {
+			logger.Object(vmi).Reason(err).Error("failed to extract key from virtualmachine.")
+			return &syncErrorImpl{fmt.Errorf("failed to extract key from virtualmachine. %v", err), FailedHotplugSyncReason}
+		}
+		if !c.requeueAfter(currentPod, key, c.requeTimeByVolumes(len(hotplugVolumes))*time.Second) {
+			if err := c.createAttachmentPodIfNotRunning(vmi, currentPod, oldPods, virtLauncherPod, readyHotplugVolumes); err != nil {
+				return err
```

Comment:
The change in the meaning of what the `currentPod` appears to make this code more difficult to follow. Due to this change, now there's if statements sprinkled around the code where the logic is not self evident (at least to me)

for example, the `deleteOldAttachmentPods` function, I see something like this below.

```
	for i, attachmentPod := range oldPods {
		if i > 0 || (currentPod != nil && currentPod.Status.Phase == k8sv1.PodRunning) {
			if attachmentPod.Status.Phase != k8sv1.PodPending {
				if err := c.deleteAttachmentPodForVolume(vmi, attachmentPod); err != nil {
					return err
				}
```

Then that logic is wrapped in a higher level cleanup function that makes this more convoluted.

```
	if c.noReadyVolumesOrCurrentPodRunning(readyHotplugVolumes, currentPod) {
		logger.V(4).Info("Deleting old attachment pods")
		// Delete old attachment pod
		if err := c.deleteOldAttachmentPods(vmi, currentPod, oldPods); err != nil {
			return &syncErrorImpl{fmt.Errorf("failed to delete pod: %v", err), FailedDeletePodReason}
		}
	}
	if c.noReadyVolumesAndCurrentPod(readyHotplugVolumes, currentPod) {
		logger.V(4).Info("Deleting current attachment pods")
		if err := c.deleteAttachmentPodForVolume(vmi, currentPod); err != nil {
			return &syncErrorImpl{fmt.Errorf("failed to delete pod: %v", err), FailedDeletePodReason}
		}
	}
```

previously this same logic for cleaning up old pods (shown below) was dead simple because currentPod was not also an oldPod too. There was a clear delineation between the thing we want to keep `currentPod` and the things we no longer need `oldPods`. 

```
		for _, attachmentPod := range oldPods {
			if err := c.deleteAttachmentPodForVolume(vmi, attachmentPod); err != nil {
				return &syncErrorImpl{fmt.Errorf("Error deleting attachment pod %v", err), 
```

A simple way to address this is to keep the previous meaning of `oldPods` and run the entire `oldPods` slice through a function to detect the most recent creation time, and ensure that time is outside of our window where we want to delay creation. something like...

```
func (c *VMIController) requeueAfter(oldPods *k8sv1.Pod, key string, requeueTime time.Duration)  bool {
	sort.Slice(oldPods, func(i, j int) bool {
		return oldPods[i].CreationTimestamp.Time.After(oldPods[j].CreationTimestamp.Time)
	})
	if len(oldPods) >0 && oldPod[0].CreationTimestamp.Time.After(time.Now().Add(-1*requeueTime)) {
		requeueTime := requeueTime-time.Since(oldPod[0].CreationTimestamp.Time))
		return true, &requeueTime
	}
	return false, nil
}
.
.
. 
        // in handleHotplugVolumes()
        if currentPod == nil && len(readyHotplugVolumes) > 0) {
               rateLimited, waitTime := requeueAfter(oldPods, time.Duration(len(readyHotPlugVolumes) / -10)
               if rateLimited {
                    c.Queue.AddAfter(key, *waitTime)
                } else if err := c.createAttachmentPod(vmi, virtLauncherPod, readyHotplugVolumes); err != nil {
			return err
		}
	}

```

Reply:
So the reason I changed the meaning of currentPod is because there was a subtle bug in the original version. If the currentPod was not yet running it would not be considered the currentPod and that value would be nil, and a new pod would be started since there was no current pod. This could lead to a large number of attachment pods being created if one added a bunch of volumes at once when no volumes existed. So I had to change the definition of the current pod to either a running pod, OR the oldest pending pod. I will see if I can simplify the logic here to make it easier to understand.

Reply:
i see. From what I understand, that primarily impacts the creation logic. Maybe a simple modification to my previous code recommendation would be

```
func hasPendingPods(pods []k8sv1.Pod) bool {
    for _, pod :=  range pods {
        if pod.Status.Phase == k8sv1.PodRunning || pod.Status.Phase == k8sv1.PodSucceeded || pod.Status.Phase == k8sv1.PodFailed {
            continue
        }
        return true
    }
    return false
}

func (c *VMIController) requeueAfter(oldPods []k8sv1.Pod, key string, requeueTime time.Duration)  bool {
	sort.Slice(oldPods, func(i, j int) bool {
		return oldPods[i].CreationTimestamp.Time.After(oldPods[j].CreationTimestamp.Time)
	})
	if len(oldPods) >0 && oldPod[0].CreationTimestamp.Time.After(time.Now().Add(-1*requeueTime)) {
		requeueTime := requeueTime-time.Since(oldPod[0].CreationTimestamp.Time))
		return true, &requeueTime
	}
	return false, nil
}
.
.
. 

        // in handleHotplugVolumes()
        if currentPod == nil && !hasPendingPods(oldPods) && len(readyHotplugVolumes) > 0) {
               rateLimited, waitTime := requeueAfter(oldPods, time.Duration(len(readyHotPlugVolumes) / -10)
               if rateLimited {
                    c.Queue.AddAfter(key, *waitTime)
                } else if err := c.createAttachmentPod(vmi, virtLauncherPod, readyHotplugVolumes); err != nil {
			return err
		}
	}
```

Reply:
So I can actually have the current pod be either a pending or a running pod, I had actually started that already just never finished it. The current code should be simpler as it doesn't have the current pod in the oldPods slice.

Reply:
I think part of the issue we're having is that we are defining the `currentPod` differently from one another.

I'm defining the `currentPod` as the pod that accurately reflects the current desired state. When we need to delete or add a new hot plug volume, and no pods accurately reflect our desired state, then there is no `currentPod`, there are only `oldPods`.

^ That was the previous intent of the code before your changes.

It appears your new definition of `currentPod` has to do what what pod is currently running, regardless if that pod matches our desired state. That leads to some complicated logic such as the code below. I want us to avoid this type of complexity

```
	for i, attachmentPod := range oldPods {
		if i > 0 || (currentPod != nil && currentPod.Status.Phase == k8sv1.PodRunning) {
			if attachmentPod.Status.Phase != k8sv1.PodPending {
				if err := c.deleteAttachmentPodForVolume(vmi, attachmentPod); err != nil {
					return err
				}
```